instrumentation/opentelemetry-instrumentation-threading/tests/test_threading.py

```diff
@@ -0,0 +1,203 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import List
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.threading import ThreadingInstrumentor
+from opentelemetry.test.test_base import TestBase
+
+
+class TestThreading(TestBase):
+    def setUp(self):
+        super().setUp()
+        self._tracer = self.tracer_provider.get_tracer(__name__)
+        self._mock_span_contexts: List[trace.SpanContext] = []
+        ThreadingInstrumentor().instrument()
+
+    def tearDown(self):
+        ThreadingInstrumentor().uninstrument()
+        super().tearDown()
+
+    def get_root_span(self):
+        return self._tracer.start_as_current_span("rootSpan")
+
+    def test_trace_context_propagation_in_thread(self):
+        self.run_threading_test(threading.Thread(target=self.fake_func))
+
+    def test_trace_context_propagation_in_timer(self):
+        self.run_threading_test(
+            threading.Timer(interval=1, function=self.fake_func)
+        )
+
+    def run_threading_test(self, thread: threading.Thread):
+        with self.get_root_span() as span:
+            expected_span_context = span.get_span_context()
+            thread.start()
+            thread.join()
+
+            # check result
+            self.assertEqual(len(self._mock_span_contexts), 1)
+            self.assert_span_context_equality(
+                self._mock_span_contexts[0], expected_span_context
+            )
+
+    def test_trace_context_propagation_in_thread_pool(self):
+        max_workers = 10
+        executor = ThreadPoolExecutor(max_workers=max_workers)
+
+        expected_span_contexts: List[trace.SpanContext] = []
+        futures_list = []
+        for num in range(max_workers):
+            with self._tracer.start_as_current_span(f"trace_{num}") as span:
+                expected_span_context = span.get_span_context()
+                expected_span_contexts.append(expected_span_context)
+                future = executor.submit(self.fake_func)
+                futures_list.append(future)
+
+        for future in as_completed(futures_list):
+            future.result()
+
+        # check result
+        self.assertEqual(len(self._mock_span_contexts), max_workers)
+        self.assertEqual(
+            len(self._mock_span_contexts), len(expected_span_contexts)
+        )
+        for index, mock_span_context in enumerate(self._mock_span_contexts):
+            self.assert_span_context_equality(
+                mock_span_context, expected_span_contexts[index]
+            )
```

Comment:
for the purposes of this testing, it would be much simpler to have the future's result be the span context inside of the submitted function (have `fake_func()` return the span context). Then the order is preserved in the results vs expected

Reply:
For `fake_func()` method has been used for the `threading.Thread`, would you mind to create another `fak_func()` method to do that or modify original method?

```python
# original
    def fake_func(self) -> trace.SpanContext:
        span_context = trace.get_current_span().get_span_context()
        self._mock_span_contexts.append(span_context)
        return span_context

# other, closure method
    def test_trace_context_propagation_in_thread_pool(self):
        def fake_func():
            return  trace.get_current_span().get_span_context()

       # ...ignore
       for num in range(max_workers):
            with self._tracer.start_as_current_span(f"trace_{num}") as span:
                future = executor.submit(fake_func))
```